Sources/SmokeDynamoDB/AWSDynamoDBCompositePrimaryKeyTable+updateItems.swift

```diff
@@ -222,6 +258,118 @@ public extension AWSDynamoDBCompositePrimaryKeyTable {
+                        return SmokeDynamoDBError.conditionalCheckFailed(partitionKey: key.partitionKey,
+                                                                         sortKey: key.sortKey,
+                                                                         message: cancellationReason.message)
+                    } else {
+                        return SmokeDynamoDBError.transactionUnknown(code: cancellationReason.code, message: cancellationReason.message)
+                    }
+                case "ItemCollectionSizeLimitExceeded":
+                    return SmokeDynamoDBError.transactionSizeExceeded(attemptedSize: entryCount,
+                                                                      maximumSize: maximumUpdatesPerTransactionStatement)
+                case "TransactionConflict":
+                    return SmokeDynamoDBError.transactionConflict(message: cancellationReason.message)
+
+                case "ProvisionedThroughputExceeded":
+                    return SmokeDynamoDBError.transactionProvisionedThroughputExceeded(message: cancellationReason.message)
+                default:
+                    return SmokeDynamoDBError.transactionUnknown(code: cancellationReason.code, message: cancellationReason.message)
+                }
+            }
+            
+            throw SmokeDynamoDBError.transactionCanceled(reasons: reasons ?? [])
+        } catch DynamoDBError.transactionConflict(let exception) {
+            let reason = SmokeDynamoDBError.transactionConflict(message: exception.message)
+            
+            throw SmokeDynamoDBError.transactionCanceled(reasons: [reason])
+        }
+    }
+    
+    private func writeChunkedItems<WriteEntryType: PolymorphicWriteEntry>(_ entries: [WriteEntryType]) async throws
+    {
+        // if there are no items, there is nothing to update
+        guard entries.count > 0 else {
+            return
+        }
+        
+        let context = StandardPolymorphicWriteEntryContext<AWSDynamoDBPolymorphicWriteEntryTransform,
+                                                           AWSDynamoDBPolymorphicTransactionConstraintTransform>(table: self)
+        let statements: [BatchStatementRequest] = try entries.map { entry -> BatchStatementRequest in
+            let transform: AWSDynamoDBPolymorphicWriteEntryTransform = try entry.handle(context: context)
+            let statement = transform.statement
+            
+            return BatchStatementRequest(consistentRead: true, statement: statement)
+        }
+        
+        let executeInput = BatchExecuteStatementInput(statements: statements)
+        
+        let response = try await dynamodb.batchExecuteStatement(input: executeInput)
+        try throwOnBatchExecuteStatementErrors(response: response)
+    }
+    
+    func bulkWrite<WriteEntryType: PolymorphicWriteEntry>(_ entries: [WriteEntryType]) async throws {
```

Comment:
Just an FYI, the problem with `monomorphicBulkWrite` and, I presume, with this one is that it's not possible to tell what statement did not execute. The reason for that is the thrown exception includes only failed items and not all items. If it were to return all items with OK/Failure status then it's possible to retry on failed items. But these response items have to match by either index or some kind of key.

Reply:
It's more of an underlying problem. It's possible to work around this for inserts. It's harder for updates because updates are not idempotent (due to list append).

Reply:
I did see that but it made sense to keep the same behaviour as `monomorphicBulkWrite`. We should address it in the next major version. I will create an issue for this.

Reply:
https://github.com/amzn/smoke-dynamodb/issues/87